shotover-proxy/src/frame/cassandra.rs

```diff
@@ -476,38 +490,262 @@ impl CassandraOperation {
+                    for operand in operands {
+                        if let Operand::Param(_) = operand {
+                            return true;
+                        }
+                    }
+                }
+            }
+            CassandraStatement::Select(select) => {
+                return CQLStatement::has_params_in_relation_elements(&select.where_clause);
+            }
+            CassandraStatement::Update(update) => {
+                for assignment_element in &update.assignments {
+                    if let Operand::Param(_) = &assignment_element.value {
+                        return true;
+                    }
+                    if let Some(assignment_operator) = &assignment_element.operator {
+                        match assignment_operator {
+                            AssignmentOperator::Plus(operand) => {
+                                if let Operand::Param(_) = operand {
+                                    return true;
+                                }
+                            }
+                            AssignmentOperator::Minus(operand) => {
+                                if let Operand::Param(_) = operand {
+                                    return true;
+                                }
+                            }
+                        }
+                    }
+                }
+                if CQLStatement::has_params_in_relation_elements(&update.where_clause) {
+                    return true;
+                }
+                if CQLStatement::has_params_in_relation_elements(&update.if_clause) {
+                    return true;
+                }
+            }
+            _ => {}
+        }
+        false
+    }
+}
+
+#[derive(PartialEq, Debug, Clone)]
+pub struct CQL {
+    pub statements: Vec<CassandraStatement>,
+}
+
+impl CQL {
+    fn to_query_string(&self) -> String {
+        self.statements.iter().map(|c| c.to_string()).join("; ")
+    }
+
+    /// the CassandraAST handles multiple queries in a string separated by semi-colons: `;` however
+    /// CQL only stores one query so this method only returns the first one if there are multiples.
+    pub fn parse_from_string(cql_query_str: &str) -> Self {
+        debug!("parse_from_string: {}", cql_query_str);
+        let ast = CassandraAST::new(cql_query_str);
+
+        let statements = if ast.has_error() {
+            vec![CassandraStatement::Unknown(cql_query_str.to_string())]
+        } else {
+            ast.statements
+                .into_iter()
+                .map(|parsed_statement| parsed_statement.statement)
+                .collect()
+        };
+        CQL { statements }
+    }
+}
+
+pub trait ToCassandraType {
+    fn as_cassandra_type(&self) -> CassandraType;
+}
+
+impl ToCassandraType for Operand {
+    fn as_cassandra_type(&self) -> CassandraType {
+        // function to convert string to CassandraType
+        let from_string_value = |value: &str| {
```

Comment:
This can be written as a regular `fn` function:
e.g.
![image](https://user-images.githubusercontent.com/5120858/168206189-31237e55-d1a3-4c5d-a4a1-60ea61e79519.png)